src/coprocessor/dag/rpn_expr/function.rs

```diff
@@ -32,6 +49,267 @@ impl<T: RpnFunction + ?Sized> RpnFunction for Box<T> {
     fn args_len(&self) -> usize {
         (**self).args_len()
     }
+
+    #[inline]
+    fn eval(
+        &self,
+        rows: usize,
+        context: &mut EvalContext,
+        payload: RpnFnCallPayload<'_>,
+    ) -> Result<VectorValue> {
+        (**self).eval(rows, context, payload)
+    }
+}
+
+pub struct Helper;
+
+impl Helper {
+    /// Evaluates a function without argument to produce a vector value.
+    ///
+    /// The function will be called multiple times to fill the vector.
+    #[inline(always)]
+    pub fn eval_0_arg<Ret, F>(
+        rows: usize,
+        mut f: F,
+        context: &mut EvalContext,
+        payload: RpnFnCallPayload<'_>,
+    ) -> Result<VectorValue>
+    where
+        Ret: Evaluable,
+        F: FnMut(&mut EvalContext, RpnFnCallPayload<'_>) -> Result<Ret>,
+    {
+        assert_eq!(payload.args_len(), 0);
+
+        let mut result = Vec::with_capacity(rows);
+        for _ in 0..rows {
+            result.push(f(context, payload)?);
+        }
+        Ok(Ret::into_vector_value(result))
+    }
+
+    /// Evaluates a function with 1 scalar or vector argument to produce a vector value.
+    ///
+    /// The function will be called multiple times to fill the vector.
+    #[inline(always)]
+    pub fn eval_1_arg<Arg0, Ret, F>(
+        rows: usize,
+        mut f: F,
+        context: &mut EvalContext,
+        payload: RpnFnCallPayload<'_>,
+    ) -> Result<VectorValue>
+    where
+        Arg0: Evaluable,
+        Ret: Evaluable,
+        F: FnMut(&mut EvalContext, RpnFnCallPayload<'_>, &Arg0) -> Result<Ret>,
+    {
+        assert_eq!(payload.args_len(), 1);
+
+        let mut result = Vec::with_capacity(rows);
+        if payload.raw_arg_at(0).is_scalar() {
+            let v = Arg0::borrow_scalar_value(payload.raw_arg_at(0).scalar_value().unwrap());
+            for _ in 0..rows {
+                result.push(f(context, payload, v)?);
+            }
+        } else {
+            let v = Arg0::borrow_vector_value(payload.raw_arg_at(0).vector_value().unwrap());
+            assert_eq!(rows, v.len());
+            for i in 0..rows {
+                result.push(f(context, payload, &v[i])?);
+            }
```

Comment:
```rust
for value in v {
    result.push(f(context, payload, value)?);
}
```

Reply:
Hmmm. I think this kind of style (including your comment in other unary / binary function) is not as easy to understand as current ones, because they use different syntax for functions serve as the same purpose? While if we access it using index, the syntax looks to be very unified.

Reply:
I don't see what this syntax is unifying with, could you elaborate? The `for` loop with iterator (including `.zip`) is used quite a lot within TiKV, so I don't think it is really harder to understand.

Reply:
The syntax now looks like:

unary function scalar value: `for _ in ... { ... f(arg0) }`
unary function vector value: `for i in ... { ... f(arg0[i]) }`
binary function scalar scalar value: `for _ in ... { ... f(arg0, arg1) }`
binary function scalar vector value: `for i in ... { ... f(arg0, arg1[i]) }`
binary function vector scalar value: `for i in ... { ... f(arg0[i], arg1) }`
binary function vector vector value: `for i in ... { ... f(arg0[i], arg1[i]) }`

which is quite easy to understand the all by only looking at one.

Reply:
I think it's better to not use `for i in 0..rows`
See https://rust-lang.github.io/rust-clippy/master/index.html#needless_range_loop

Reply:
Actually we disabled this clippy rule for the reason that this rule does not always produce more readable code :)